File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def dfs(v, graph, visited):
    visited[v] = True
    count = 1  # 현재 노드 카운트
    for u in graph[v]:
        if not visited[u]:
            count += dfs(u, graph, visited)
    return count

def solution(n, wires):
    graph = [[] for _ in range(n + 1)]
    for v, u in wires:
        graph[v].append(u)
        graph[u].append(v)

    answer = 100
    for i in range(n - 1):
        visited = [False for _ in range(n + 1)]
        v1, v2 = wires[i]
        visited[v2] = True
        tmp = abs(dfs(v1, graph, visited) - dfs(v2, graph, visited))
        answer = min(tmp, answer)

    return answer

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    public int dfs(int v, List<Integer>[] graph, boolean[] visited) {
        visited[v] = true;
        int cnt = 1;
        for (int i = 0; i < graph[v].size(); i++) {
            int u = graph[v].get(i);
            if (!visited[u]) {
                cnt += dfs(u, graph, visited);
            }
        }
        return cnt;
    }
    
    public int solution(int n, int[][] wires) {
        List<Integer>[] graph = new ArrayList[n + 1];
        
        for (int i = 0; i < n + 1; i++) {
            graph[i] = new ArrayList<>();        
        }
        
        for (int i = 0; i < wires.length; i++) {
            int a = wires[i][0];
            int b = wires[i][1];
            graph[a].add(b);
            graph[b].add(a);
        }
        
        int answer = 100;
        for (int i = 0; i < n-1; i++) {
            boolean[] visited = new boolean[n+1];
            int a = wires[i][0];
            int b = wires[i][1];
            visited[a] = true;
            int temp = Math.abs(dfs(b, graph, visited) - dfs(a, graph, visited));
            answer = Math.min(temp, answer);
        }
        return answer;
    }
    
    
}